# PRÁCTICA 2 - FAA
### K-NN y K-MEANS
#### 10 Noviembre 2022
Alejandro Coloma y S. Xiao Férnandez - Grupo 9

### K-NN propio

#### Resultados

In [1]:
from MainPractica2FAA import *
test_propio()

0.19607843137254902 0.0160097368809358
0.21960784313725487 0.03756890944627429
0.1843137254901961 0.015686274509803918
0.22222222222222224 0.021077798034767444
0.22222222222222224 0.026143790849673203
0.20784313725490197 0.025615579009497288
0.20915032679738563 0.022260635772452807
0.2261437908496732 0.04227702386613393
0.19084967320261437 0.019996154955265827
0.21568627450980396 0.03743221193013426
0.2143790849673203 0.03492912213303893
0.2065359477124183 0.028519508796433234
0.017699115044247787 0.01119390322183497
0.02831858407079646 0.017159928698818862
0.023008849557522124 0.015429730773595305
0.038938053097345125 0.016413484062824255
0.046017699115044254 0.018049626596788618
0.04070796460176991 0.015429730773595303
0.03893805309734513 0.01437882903475391
0.053097345132743355 0.012515164268788452
0.049557522123893805 0.012004123863938527
0.06194690265486726 0.017699115044247787
0.0654867256637168 0.012004123863938529
0.06548672566371681 0.022803714560575443
               diabetes

diabetes - Media diabetes - D. Típica wdbc - Media  \
Normalizado #K                                                      
True        5        0.19607843           0.01600974   0.01769912   
            7        0.21960784           0.03756891   0.02831858   
            10       0.18431373           0.01568627   0.02300885   
            15       0.22222222           0.02107780   0.03893805   
            17       0.22222222           0.02614379   0.04601770   
            20       0.20784314           0.02561558   0.04070796   
False       5        0.20915033           0.02226064   0.03893805   
            7        0.22614379           0.04227702   0.05309735   
            10       0.19084967           0.01999615   0.04955752   
            15       0.21568627           0.03743221   0.06194690   
            17       0.21437908           0.03492912   0.06548673   
            20       0.20653595           0.02851951   0.06548673   

               wdbc - D. Típica  
Normalizado #K                   
True        5        0.01119390  
            7        0.01715993  
            10       0.01542973  
            15       0.01641348  
            17       0.01804963  
            20       0.01542973  
False       5        0.01437883  
            7        0.01251516  
            10       0.01200412  
            15       0.01769912  
            17       0.01200412  
            20       0.02280371

#### Explicación del código y decisiones tomadas.

En el entrenamiento, tampoco hemos hecho mucho, ya que al ser un metodo retardado, esta funcion es muy limitada. Nos hemos centrado en poner la normalización y en el calculo de las medias y las desviaciones tipicas para la normalización.

Para la clasificación, hemos calculado el círculo de los K vecinos y hemos estimado como clase prevista la clase mayoritaria.

La distancia que hemos usado ha sido la euclidea, y en el test, hemos decidido hacerlo con tan solo validación cruzada, ya que en la anterior práctica era la que menos media de error nos daba. Hemos decidido hacerlo solo con VC ya que si teníamos que hacerlo con VS también, nos iba a tardar mucho tiempo.

#### Análisis de los resultados
**Obtener los resultados tanto para datos normalizados como sin normalizar, con el objetivo de justificar el rendimiento del algoritmo en base a estas características. Analizar el impacto de K.** 

Las K elegidas han sido 5, 7, 10, 15, 17, 20. De esta manera podemos ir viendo el rendimiento del algoritmo y como este se comporta en base a los valores que pongamos de vecindad.

Además de esto, vemos como los datos también están normalizados, para evitar problemas de escala.

Mirando las tablas podemos observar que el valor óptimo para el dataset diabetes, es k=10 sin normalizar, porque nos da un error del 20.6%, que tampoco difiere mucho del dataset normlaizado, un 20.9% de media de error.

Sin embargo, para el dataset wdbc, el valor óptimo de K es 5 cuando el dataset está normalizado. Podemos afirmar esto ya que nos da una media de error mas baja que con otros K y la desviación tipica es menor. En el caso donde no está normalizado, el mejor casos igue siendo K=5 pero la media de error es casi un 1% mas alta.

### K-means

In [2]:
from MainPractica2FAA import *
test_KMeans()

Para K=3, numero de iteraciones: 6
	50 puntos pertenecen al cluster 2 con centroide [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
	62 puntos pertenecen al cluster 0 con centroide [5.901612903225807, 2.748387096774194, 4.393548387096775, 1.4338709677419357]
	38 puntos pertenecen al cluster 1 con centroide [6.8500000000000005, 3.073684210526315, 5.742105263157893, 2.0710526315789473]
[[ 0  0 50]
 [48  2  0]
 [14 36  0]]
Para K=1, numero de iteraciones: 1
	150 puntos pertenecen al cluster 0 con centroide [5.843333333333335, 3.0540000000000007, 3.7586666666666693, 1.1986666666666672]
Para K=2, numero de iteraciones: 2
	53 puntos pertenecen al cluster 1 con centroide [5.005660377358491, 3.360377358490567, 1.562264150943396, 0.2886792452830188]
	97 puntos pertenecen al cluster 0 con centroide [6.30103092783505, 2.8865979381443303, 4.958762886597939, 1.6958762886597945]
Para K=4, numero de iteraciones: 11
	50 puntos pertenecen al cluster 2 con centroide [5.005999999999999


#### Explicación del código y decisiones tomadas.

Primero seleccionamos aleatoriamente k patrones para que sean los centroides de los clústers. Tras esto asignamos cada uno del resto de patrones al centroide más cercano.

Una vez hecho esto, en cada clúster elegir un nuevo centroide de entre los patrones miembros del cluster, usando el centro de masas. 

Si vemos que los clusters no han cambiado de una iteracion a otra, es que ya ha terminado de clasificar y se sale del bucle.

#### Análisis de los resultados
**Para K=3, comprobar si se puede asignar de forma unívoca cada clúster a un tipo de flor atendiendo a la clase mayoritaria de los patrones agrupados por clúster**

Como vemos en los resultados devueltos, vemos que divide los datos de una forma equitativa, por lo que si que los está ordenando bien, ya que al ver el archivo original, observamos como hay una division de las flores equitativa. Viendo los puntos que coje cada cluster podemos deducir a que flor pertenece cada uno.

**Probar con otros valores de K=[1,2,4,5] y comparar los resultados con los obtenidos para K=3. ¿Cuál sería el mejor K para este problema?**

Varía con cada ejecucion, pero podemos observar que como mucho hace 5 iteraciones. Todo depende de los clustyers que quieras tener. Vemos que cuantos mas clusters tengamos, menos equitativos estan repartidos los puntos. Es decir, al haber menos clusters (2 o 3), los numeros de putos en cada clusters son mas parecidos entre si.

Cuanto mas alejados estan, (4 y 5), mas diferencias entre las plantas encuentra, por lo que la clasificacion desvaría mas de lo que debería ser.

Este problema se resuelve bien con K=3, ya que hay 3 clases de flores.

**Analizar si cambiarían los resultados normalizando los datos**

Si que cambiarian los datos. Al normalizar esperamos que las medias mejoren, aunque en este caso, tampoco creemos que mejorasen mucho porque la normalizacion consiste en transformar los datos en una escala "que se calcula del cociente del valor en cuestión menos el valor mínimo de la lista entre el rango de la lista". En este caso, al tener todos los valores mas o menos la misma escala, no notariamos mucha diferencia.

**Analizar qué tipos de flor se identifican más fácilmente y cuales se confunden entre ellos. Para ello se puede hacer uso de la matriz de confusión multiclase a partir del número de puntos en cada clase y sabiendo que hay 50 ejemplos de cada tipo de flor**

La matriz se nos presenta de forma no pareja, ya que cada vez se ejecua, los verdaderos positivos se colocan en cualquier columna. Estamos seguros de que cada fila representa un cluster, pero como las llaves del diccionario no tienen por que ser 0, 1 o 2, se puede llegar a obtener otro orden. En este caso inuimos que en el cluster 0 hay 50 vrdaderos positivos, en el cluster 1, 47 verdaderos positivos y 3 falsos positivios. Por ultimo, en el cluster 2, 36 verdaderos positivos y 14 falsos positivos.

### Knn & Kmeans SKLearn

In [1]:
from MainPractica2FAA import *
test_SKL()

diabetes - Media diabetes - D. Típica wdbc - Media  \
Normalizado #K                                                      
True        5        0.29297173           0.02325546   0.18632200   
            7        0.34373143           0.03909551   0.17761217   
            10       0.31383584           0.02452231   0.15632666   
            15       0.29165606           0.01429015   0.14934016   
            17       0.29816654           0.03848643   0.16697718   
            20       0.28517953           0.02064723   0.16360814   
False       5        0.27867753           0.04011015   0.07031517   
            7        0.28249724           0.01876448   0.06682192   
            10       0.27730244           0.02153314   0.06854526   
            15       0.27083439           0.01209387   0.06675982   
            17       0.24609965           0.02549873   0.06683745   
            20       0.26572447           0.03410164   0.07208508   

               wdbc - D. Típica  
Normalizado #K                   
True        5        0.03874888  
            7        0.05176111  
            10       0.04765623  
            15       0.04662657  
            17       0.01936950  
            20       0.05583673  
False       5        0.02006621  
            7        0.01647211  
            10       0.00660210  
            15       0.01416190  
            17       0.02143998  
            20       0.01715573

#### Explicación del código y decisiones tomadas.

Para normalizar utilizamos StandardScaler ya que otra posibilidad habría sido usar prepocessing, pero solo normaliza la media y para kmeans no es suficiente solo con eso. Se necesita tambien tener en cuenta la varienza

En cuanto a Knn de SKL 

n_neighbors=k: numero de vecinos

metric='minkowski' y p=2, que nos indica que trabajamos con la distancia euclidea

#### Análisis de los resultados

El valor óptimo para el dataset diabetes, es k=12 sin normalizar, porque nos da una media de error mas baja que las otras. Sin embargo, para el dataset wdbc, el valor óptimo de K es 17, también si n normlaizar. Vemos como normalizar los datos empeora el error. La diferencia se puede ver sobre todo en wdbc. Cosa que nos resultó raro al verlo por primera vez, ya que normalizado deberia dar un error mas bajo.


In [4]:
from MainPractica2FAA import *
test_KmeansSLK()

Para K=3 clusters, se han encontrado
	Centroide 0: [6.62244898 2.98367347 5.57346939 2.03265306 2.        ] con 49 puntos pertenecientes
	Centroide 1: [5.006 3.418 1.464 0.244 0.   ] con 50 puntos pertenecientes
	Centroide 2: [5.91568627 2.76470588 4.26470588 1.33333333 1.01960784] con 51 puntos pertenecientes
Para K=1 clusters, se han encontrado
	Centroide 0: [5.84333333 3.054      3.75866667 1.19866667 1.        ] con 150 puntos pertenecientes
Para K=2 clusters, se han encontrado
	Centroide 0: [6.27373737 2.87575758 4.92525253 1.68181818 1.50505051] con 99 puntos pertenecientes
	Centroide 1: [5.00784314 3.4        1.49411765 0.26078431 0.01960784] con 51 puntos pertenecientes
Para K=4 clusters, se han encontrado
	Centroide 0: [5.006 3.418 1.464 0.244 0.   ] con 50 puntos pertenecientes
	Centroide 1: [5.87959184 2.75306122 4.23673469 1.32244898 1.02040816] con 49 puntos pertenecientes
	Centroide 2: [7.38571429 3.13571429 6.22857143 2.08571429 2.        ] con 14 puntos pertenecientes
	

#### Explicación del código y decisiones tomadas.

Para normalizar utilizamos StandardScaler ya que otra posibilidad habría sido usar prepocessing, pero solo normaliza la media y para kmeans no es suficiente solo con eso. Se necesita tambien tener en cuenta la varienza

En cuanto a Knn de SKL 

n_neighbors=k: numero de vecinos

metric='minkowski' y p=2, que nos indica que trabajamos con la distancia euclidea

#### Análisis de los resultados

El valor óptimo para el dataset diabetes, es k=12 sin normalizar, porque nos da una media de error mas baja que las otras. Sin embargo, para el dataset wdbc, el valor óptimo de K es 17.

Vemos como normalizar los datos empeora el error. La diferencia se puede ver sobre todo en wdbc. 


#### Explicación del código y decisiones tomadas.

Para Kmeans en SKLearn, hemos puesto los parametros 

n_clusters=num_clusters

init='k-means++': Los centroides iniciales no son completamente aleatorios

tol=0.0001: La tolerancia a que los centroides sean iguales. Es decir, en nuestro Kmeans, se para cuando los centroides son iguales. Aqui se puede poner una tolerancia.                                

#### Análisis de los resultados

Vemos como en esta implementacion los puntos (en k=3) estan mas repartidos equitativamente. Sabemos esto porque en el dataset, tenemos 50 flores de cada timpo y aqui nos clasifica 51, 50 y 49, en nuestra implementacion, 50, 39, 61.

### Conclusión
**Comparar y analizar los resultados propios con los de Scikit-Learn**

Para K-nn, sale menor media de error en nuestra implementacion, aunque tampoco se ve mucha diferencia. Podemos encontrar mas diferencia entre los tiempor de ejecucion. SKL ha tardado 0.8s y nuestra implementacion de Knn ha tardado 13min 40.0s.

En cuanto a Kmeans, ya hemos explicado anteriormente que es lo que pasaba, donde la implementacion de SKL los clasifica con mas precisión.